src/main/java/org/terasology/weatherManager/systems/BlockPlacingWeatherSystem.java

```diff
@@ -72,67 +76,102 @@ public void onPlaceEvent(PeriodicActionTriggeredEvent event, EntityRef worldEnti
         int iter = 0;
-        boolean lastGround = false;
-        boolean placed = false;
-        while (!placed || iter < SNOW_BLOCK_RANGE * 2) {
-            Block current = worldProvider.getBlock(x, currentY, z);
-            if (current.equals(air) && lastGround) {
-                worldProvider.setBlock(new Vector3i(x, currentY, z), snow);
-                placed = true;
-            } else if (current.equals(air)) {
-                currentY--;
-                lastGround = false;
-            } else if (current.isPenetrable() || !current.isAttachmentAllowed()) {
-                break;
-            } else if (!current.equals(snow)) {
-                lastGround = true;
-                currentY++;
-            } else {
-                placed = true; //break out to avoid double-placing snow
-            }
+        while (iter < SNOW_BLOCK_RANGE * 2 && worldProvider.getBlock(x, currentY, z).equals(air)) {
             iter++;
+            currentY--;
+        }
+        while (iter < SNOW_BLOCK_RANGE * 2 && !worldProvider.getBlock(x, currentY, z).equals(air)) {
+            iter++;
+            currentY++;
+        }
+        if (iter >= SNOW_BLOCK_RANGE * 2) {
+            return null;
+        }
+
+        if (worldProvider.getSunlight(x, currentY, z) != ChunkConstants.MAX_SUNLIGHT) {
+            // The block isn't actually exposed to the weather.
+            return null;
+        }
+        Block ground = worldProvider.getBlock(x, currentY-1, z);
+        if (ground.equals(toCheck)) {
+            return new Vector3i(x, currentY-1, z);
+        } else if (toCheck.equals(air) && !ground.isPenetrable() && ground.isAttachmentAllowed()) {
+            return new Vector3i(x, currentY, z);
+        } else {
+            return null;
         }
     }
 
-    private void removeSnow(Vector3i playerPos) {
-        FastRandom rand = new FastRandom();
-        int x = (int) playerPos.x + rand.nextInt(SNOW_BLOCK_RANGE * 2) - SNOW_BLOCK_RANGE;
-        int z = (int) playerPos.z + rand.nextInt(SNOW_BLOCK_RANGE * 2) - SNOW_BLOCK_RANGE;
-        int currentY = (int) playerPos.y + SNOW_BLOCK_RANGE;
-        int iter = 0;
-        boolean lastGround = false;
-        boolean placed = false;
-        while (!placed || iter < SNOW_BLOCK_RANGE * 2) {
-            Block current = worldProvider.getBlock(x, currentY, z);
-            if (current.equals(snow)) {
-                worldProvider.setBlock(new Vector3i(x, currentY, z), air);
-                placed = true;
-            } else if (lastGround) {
-                placed = true; //break out if there is no snow
-            } else if (current.equals(air)) {
-                currentY--;
-                lastGround = false;
-            } else {
-                lastGround = true;
-                currentY++;
-            }
-            iter++;
+    private void placeSnow(Vector3i playerPos) {
+        int x = getValueToPlaceBlock(playerPos.x);
+        int z = getValueToPlaceBlock(playerPos.x);
```

Comment:
```suggestion
        int z = getValueToPlaceBlock(playerPos.z);
```

Also change this in the other methods.